File: omega/core/deterministic_compiler.py
"""
omega.core.deterministic_compiler
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
DeterministicCompiler — freezes a node's current parameters into a versioned,
LLM-free strategy that can execute with zero external calls.

This is the PICO-mode execution engine. When a node is at AutonomyLevel.PICO,
the controller compiles its current parameters into a DeterministicStrategy.
The strategy executes rule-based logic only — no brain, no LLM, no network.

Design
------
DeterministicStrategy   — frozen, versioned rule set + execute()
DeterministicCompiler   — compiles a node snapshot → DeterministicStrategy
Signal                  — output of DeterministicStrategy.execute()
StrategyRegistry        — in-memory + optional StateStore-backed version history

Rollback
--------
Each compilation is assigned a monotonically increasing version. Any prior
version can be retrieved and re-activated via rollback().

Usage
-----
    compiler = DeterministicCompiler()
    strategy = compiler.compile_strategy(node)
    signal = strategy.execute(market_data)
    print(signal.direction, signal.confidence)

    # Rollback to a previous version
    old_strategy = compiler.rollback(node_id="my-node", version=1)
"""

from __future__ import annotations

import copy
import hashlib
import json
import time
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

# ---------------------------------------------------------------------------
# Signal — output of a deterministic strategy
# ---------------------------------------------------------------------------


class SignalDirection(str, Enum):  # noqa: UP042
    LONG = "long"
    SHORT = "short"
    FLAT = "flat"  # no position


@dataclass
class Signal:
    """
    Trading / action signal produced by a DeterministicStrategy.

    direction   — LONG, SHORT, or FLAT
    confidence  — [0.0, 1.0]; how strongly the rules fired
    metadata    — any additional context (rule names, thresholds used, etc.)
    generated_at — Unix timestamp of signal creation
    """

    direction: SignalDirection = SignalDirection.FLAT
    confidence: float = 0.0
    metadata: dict[str, Any] = field(default_factory=dict)
    generated_at: float = field(default_factory=time.time)

    def is_actionable(self, min_confidence: float = 0.5) -> bool:
        return self.direction != SignalDirection.FLAT and self.confidence >= min_confidence


# ---------------------------------------------------------------------------
# DeterministicStrategy — frozen, versioned rule set
# ---------------------------------------------------------------------------


@dataclass
class DeterministicStrategy:
    """
    A frozen, versioned representation of a node's rule-based logic.

    Created by DeterministicCompiler.compile_strategy().  Executes with
    zero LLM calls — purely deterministic given the frozen parameters.

    Attributes
    ----------
    strategy_id     — unique ID for this compiled artifact
    node_id         — source node
    node_version    — node version at compile time
    version         — monotonically increasing compilation version
    parameters      — deep-frozen copy of node parameters at compile time
    rules           — ordered list of rule dicts (see _apply_rules)
    compiled_at     — Unix timestamp
    checksum        — SHA-256 of the serialised parameters (integrity check)
    """

    strategy_id: str
    node_id: str
    node_version: str
    version: int
    parameters: dict[str, Any]
    rules: list[dict[str, Any]]
    compiled_at: float = field(default_factory=time.time)
    checksum: str = ""

    # ------------------------------------------------------------------
    # Execution — pure rule engine, no LLM
    # ------------------------------------------------------------------

    def execute(self, market_data: dict[str, Any]) -> Signal:
        """
        Evaluate frozen rules against *market_data* and return a Signal.

        Rules are evaluated in order; the first rule whose conditions all
        pass determines the signal direction. If no rule fires, returns
        Signal(FLAT, 0.0).

        Each rule dict has the shape:
        {
            "name":       str,
            "direction":  "long" | "short" | "flat",
            "conditions": [
                {"field": str, "op": ">"|"<"|">="|"<="|"=="|"!=", "value": float}
            ],
            "confidence": float
        }
        """
        for rule in self.rules:
            if self._evaluate_rule(rule, market_data):
                return Signal(
                    direction=SignalDirection(rule.get("direction", "flat")),
                    confidence=float(rule.get("confidence", 0.5)),
                    metadata={
                        "rule_name": rule.get("name", "unnamed"),
                        "strategy_id": self.strategy_id,
                        "strategy_version": self.version,
                    },
                )
        return Signal(
            direction=SignalDirection.FLAT,
            confidence=0.0,
            metadata={"strategy_id": self.strategy_id, "reason": "no_rule_fired"},
        )

    def _evaluate_rule(self, rule: dict[str, Any], data: dict[str, Any]) -> bool:
        """Return True if all conditions in *rule* pass against *data*."""
        conditions = rule.get("conditions", [])
        if not conditions:
            return False
        for cond in conditions:
            field_val = self._resolve_field(cond.get("field", ""), data)
            if field_val is None:
                return False
            op = cond.get("op", "==")
            threshold = cond.get("value", 0.0)
            if not self._compare(field_val, op, threshold):
                return False
        return True

    @staticmethod
    def _resolve_field(field_path: str, data: dict[str, Any]) -> Any:
        """Support dot-notation field paths like 'price.close'."""
        parts = field_path.split(".")
        current: Any = data
        for part in parts:
            if not isinstance(current, dict):
                return None
            current = current.get(part)
            if current is None:
                return None
        return current

    @staticmethod
    def _compare(value: Any, op: str, threshold: float) -> bool:
        try:
            v = float(value)
        except (TypeError, ValueError):
            return False
        ops = {
            ">": v > threshold,
            "<": v < threshold,
            ">=": v >= threshold,
            "<=": v <= threshold,
            "==": v == threshold,
            "!=": v != threshold,
        }
        return ops.get(op, False)

    # ------------------------------------------------------------------
    # Integrity
    # ------------------------------------------------------------------

    def verify_checksum(self) -> bool:
        """Return True if the parameters have not been tampered with."""
        expected = _compute_checksum(self.parameters)
        return expected == self.checksum

    # ------------------------------------------------------------------
    # Serialisation
    # ------------------------------------------------------------------

    def to_dict(self) -> dict[str, Any]:
        return {
            "strategy_id": self.strategy_id,
            "node_id": self.node_id,
            "node_version": self.node_version,
            "version": self.version,
            "parameters": self.parameters,
            "rules": self.rules,
            "compiled_at": self.compiled_at,
            "checksum": self.checksum,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> DeterministicStrategy:
        return cls(
            strategy_id=d["strategy_id"],
            node_id=d["node_id"],
            node_version=d["node_version"],
            version=int(d["version"]),
            parameters=d["parameters"],
            rules=d["rules"],
            compiled_at=float(d.get("compiled_at", time.time())),
            checksum=d.get("checksum", ""),
        )


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _compute_checksum(parameters: dict[str, Any]) -> str:
    serialised = json.dumps(parameters, sort_keys=True, default=str)
    return hashlib.sha256(serialised.encode()).hexdigest()


def _extract_parameters(node: Any) -> dict[str, Any]:
    """
    Best-effort extraction of a node's current parameters.

    Tries, in order:
    1. node.get_parameters()
    2. node.parameters (attribute)
    3. node.get_state().metadata
    4. empty dict
    """
    if hasattr(node, "get_parameters") and callable(node.get_parameters):
        try:
            return copy.deepcopy(node.get_parameters())
        except Exception:
            pass
    if hasattr(node, "parameters") and isinstance(node.parameters, dict):
        return copy.deepcopy(node.parameters)
    if hasattr(node, "get_state"):
        try:
            state = node.get_state()
            return copy.deepcopy(state.metadata)
        except Exception:
            pass
    return {}


def _extract_rules(node: Any, parameters: dict[str, Any]) -> list[dict[str, Any]]:
    """
    Extract or synthesise rules from the node's current configuration.

    Tries node.get_rules() first. Falls back to synthesising rules from
    common parameter keys (e.g. 'signal_threshold', 'rsi_oversold', etc.)
    so that even nodes without explicit rule support get a sensible default.
    """
    if hasattr(node, "get_rules") and callable(node.get_rules):
        try:
            node_rules: Any = node.get_rules()
            if isinstance(node_rules, list) and node_rules:  # only use if non-empty
                return copy.deepcopy(node_rules)
        except Exception:
            pass

    # Synthesise rules from well-known parameter keys
    rules: list[dict[str, Any]] = []

    # RSI-style rules
    if "rsi_oversold" in parameters:
        rules.append(
            {
                "name": "rsi_oversold_long",
                "direction": "long",
                "conditions": [
                    {"field": "rsi", "op": "<=", "value": float(parameters["rsi_oversold"])}
                ],
                "confidence": 0.7,
            }
        )
    if "rsi_overbought" in parameters:
        rules.append(
            {
                "name": "rsi_overbought_short",
                "direction": "short",
                "conditions": [
                    {"field": "rsi", "op": ">=", "value": float(parameters["rsi_overbought"])}
                ],
                "confidence": 0.7,
            }
        )

    # Generic threshold rules
    if "signal_threshold_long" in parameters:
        rules.append(
            {
                "name": "signal_long",
                "direction": "long",
                "conditions": [
                    {
                        "field": "signal",
                        "op": ">=",
                        "value": float(parameters["signal_threshold_long"]),
                    }
                ],
                "confidence": float(parameters.get("signal_confidence", 0.6)),
            }
        )
    if "signal_threshold_short" in parameters:
        rules.append(
            {
                "name": "signal_short",
                "direction": "short",
                "conditions": [
                    {
                        "field": "signal",
                        "op": "<=",
                        "value": float(parameters["signal_threshold_short"]),
                    }
                ],
                "confidence": float(parameters.get("signal_confidence", 0.6)),
            }
        )

    # Always add a flat fallback (lowest priority, caught by default in execute())
    return rules


# ---------------------------------------------------------------------------
# StrategyRegistry — in-memory version history with optional persistence
# ---------------------------------------------------------------------------


class StrategyRegistry:
    """
    Versioned strategy store.

    Keeps all compiled strategies for each node indexed by version number.
    The current (latest) strategy is the one with the highest version.
    """

    def __init__(self) -> None:
        # node_id → {version: DeterministicStrategy}
        self._strategies: dict[str, dict[int, DeterministicStrategy]] = {}

    def register(self, strategy: DeterministicStrategy) -> None:
        node_strategies = self._strategies.setdefault(strategy.node_id, {})
        node_strategies[strategy.version] = strategy

    def get(self, node_id: str, version: int) -> DeterministicStrategy | None:
        return self._strategies.get(node_id, {}).get(version)

    def get_current(self, node_id: str) -> DeterministicStrategy | None:
        node_strategies = self._strategies.get(node_id, {})
        if not node_strategies:
            return None
        latest_version = max(node_strategies.keys())
        return node_strategies[latest_version]

    def list_versions(self, node_id: str) -> list[int]:
        return sorted(self._strategies.get(node_id, {}).keys())

    def next_version(self, node_id: str) -> int:
        versions = self.list_versions(node_id)
        return (max(versions) + 1) if versions else 1


# ---------------------------------------------------------------------------
# DeterministicCompiler
# ---------------------------------------------------------------------------


class DeterministicCompiler:
    """
    Compiles a node's current parameters into a DeterministicStrategy.

    Each call to compile_strategy() freezes the node's parameters as of
    that moment, assigns a new version, and registers it in the StrategyRegistry.

    Rollback
    --------
    Any previously compiled version can be retrieved with rollback(). The
    compiler does not automatically apply rolled-back strategies to nodes —
    callers must do that explicitly (e.g. by calling node.load_strategy()).

    Usage
    -----
        compiler = DeterministicCompiler()
        strategy = compiler.compile_strategy(node)
        signal = strategy.execute({"rsi": 28.0, "price": {"close": 100.0}})

        # List versions for a node
        compiler.registry.list_versions("my-node-id")

        # Rollback to version 1
        old = compiler.rollback("my-node-id", version=1)
    """

    def __init__(self, registry: StrategyRegistry | None = None) -> None:
        self.registry = registry or StrategyRegistry()

    def compile_strategy(self, node: Any) -> DeterministicStrategy:
        """
        Freeze *node*'s current parameters as a DeterministicStrategy.

        1. Extract parameters from the node (deep copy — no references)
        2. Extract / synthesise rules from the parameters
        3. Assign a monotonically increasing version
        4. Compute an integrity checksum
        5. Register and return

        The node must implement the Node base class interface (get_state()).
        """
        # Extract node identity
        try:
            state = node.get_state()
            node_id = state.node_id
            node_version = state.version
        except Exception:
            node_id = getattr(node, "node_id", str(uuid.uuid4()))
            node_version = getattr(node, "version", "unknown")

        parameters = _extract_parameters(node)
        rules = _extract_rules(node, parameters)
        version = self.registry.next_version(node_id)
        checksum = _compute_checksum(parameters)

        strategy = DeterministicStrategy(
            strategy_id=str(uuid.uuid4()),
            node_id=node_id,
            node_version=node_version,
            version=version,
            parameters=parameters,
            rules=rules,
            compiled_at=time.time(),
            checksum=checksum,
        )
        self.registry.register(strategy)
        return strategy

    def rollback(self, node_id: str, version: int) -> DeterministicStrategy | None:
        """
        Retrieve a previously compiled strategy by version number.

        Returns None if the version does not exist in the registry.
        Does NOT automatically apply the strategy to the node.
        """
        return self.registry.get(node_id, version)

    def get_current(self, node_id: str) -> DeterministicStrategy | None:
        """Return the most recently compiled strategy for a node."""
        return self.registry.get_current(node_id)

File: internal/core/deterministic_compiler.go
package core

// deterministic_compiler.go — DeterministicCompiler, DeterministicStrategy,
// Signal, StrategyRegistry.
//
// Compiles a node's current parameters into a versioned, LLM-free rule set
// (PICO-mode execution engine).  Pure deterministic logic — no network calls.

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
)

// ---------------------------------------------------------------------------
// SignalDirection
// ---------------------------------------------------------------------------

// SignalDirection represents the direction of a trading/action signal.
type SignalDirection string

const (
	DirectionLong  SignalDirection = "long"
	DirectionShort SignalDirection = "short"
	DirectionFlat  SignalDirection = "flat"
)

// ---------------------------------------------------------------------------
// Signal
// ---------------------------------------------------------------------------

// Signal is the output of a DeterministicStrategy.Execute() call.
type Signal struct {
	Direction   SignalDirection
	Confidence  float64
	Metadata    map[string]any
	GeneratedAt time.Time
}

// IsActionable returns true if the signal has a non-flat direction and meets
// the minimum confidence threshold.
func (s Signal) IsActionable(minConfidence float64) bool {
	return s.Direction != DirectionFlat && s.Confidence >= minConfidence
}

// ---------------------------------------------------------------------------
// Rule condition helpers
// ---------------------------------------------------------------------------

// RuleCondition is a single condition in a strategy rule.
type RuleCondition struct {
	Field string  // dot-notation field path in market data, e.g. "price.close"
	Op    string  // ">", "<", ">=", "<=", "==", "!="
	Value float64
}

// StrategyRule is a single rule in a DeterministicStrategy.
type StrategyRule struct {
	Name       string
	Direction  SignalDirection
	Conditions []RuleCondition
	Confidence float64
}

// evaluate returns true if all conditions pass against data.
func (r StrategyRule) evaluate(data map[string]any) bool {
	if len(r.Conditions) == 0 {
		return false
	}
	for _, c := range r.Conditions {
		v, ok := resolveField(c.Field, data)
		if !ok {
			return false
		}
		if !compareFloat(v, c.Op, c.Value) {
			return false
		}
	}
	return true
}

func resolveField(path string, data map[string]any) (float64, bool) {
	parts := strings.Split(path, ".")
	var cur any = data
	for _, p := range parts {
		m, ok := cur.(map[string]any)
		if !ok {
			return 0, false
		}
		cur = m[p]
		if cur == nil {
			return 0, false
		}
	}
	switch v := cur.(type) {
	case float64:
		return v, true
	case float32:
		return float64(v), true
	case int:
		return float64(v), true
	case int64:
		return float64(v), true
	default:
		return 0, false
	}
}

func compareFloat(v float64, op string, threshold float64) bool {
	switch op {
	case ">":
		return v > threshold
	case "<":
		return v < threshold
	case ">=":
		return v >= threshold
	case "<=":
		return v <= threshold
	case "==":
		return v == threshold
	case "!=":
		return v != threshold
	}
	return false
}

// ---------------------------------------------------------------------------
// DeterministicStrategy
// ---------------------------------------------------------------------------

// DeterministicStrategy is a frozen, versioned rule set compiled from a node's
// parameters at a specific point in time.
type DeterministicStrategy struct {
	StrategyID  string
	NodeID      string
	NodeVersion string
	Version     int
	Parameters  map[string]any
	Rules       []StrategyRule
	CompiledAt  time.Time
	Checksum    string // SHA-256 of serialised parameters
}

// Execute evaluates the frozen rules against marketData and returns a Signal.
// Rules are evaluated in order; the first matching rule wins.  If no rule
// fires, a flat Signal with zero confidence is returned.
func (s *DeterministicStrategy) Execute(marketData map[string]any) Signal {
	for _, rule := range s.Rules {
		if rule.evaluate(marketData) {
			return Signal{
				Direction:  rule.Direction,
				Confidence: rule.Confidence,
				Metadata: map[string]any{
					"rule_name":        rule.Name,
					"strategy_id":      s.StrategyID,
					"strategy_version": s.Version,
				},
				GeneratedAt: time.Now(),
			}
		}
	}
	return Signal{
		Direction:  DirectionFlat,
		Confidence: 0,
		Metadata:   map[string]any{"strategy_id": s.StrategyID, "reason": "no_rule_fired"},
		GeneratedAt: time.Now(),
	}
}

// VerifyChecksum returns true if the parameters have not been tampered with.
func (s *DeterministicStrategy) VerifyChecksum() bool {
	return computeChecksum(s.Parameters) == s.Checksum
}

// ToMap serialises the strategy to a plain map for persistence.
func (s *DeterministicStrategy) ToMap() map[string]any {
	rules := make([]map[string]any, len(s.Rules))
	for i, r := range s.Rules {
		conds := make([]map[string]any, len(r.Conditions))
		for j, c := range r.Conditions {
			conds[j] = map[string]any{"field": c.Field, "op": c.Op, "value": c.Value}
		}
		rules[i] = map[string]any{
			"name":       r.Name,
			"direction":  string(r.Direction),
			"conditions": conds,
			"confidence": r.Confidence,
		}
	}
	return map[string]any{
		"strategy_id":  s.StrategyID,
		"node_id":      s.NodeID,
		"node_version": s.NodeVersion,
		"version":      s.Version,
		"parameters":   s.Parameters,
		"rules":        rules,
		"compiled_at":  s.CompiledAt.Unix(),
		"checksum":     s.Checksum,
	}
}

// ---------------------------------------------------------------------------
// StrategyRegistry
// ---------------------------------------------------------------------------

// StrategyRegistry stores all compiled strategies for each node, indexed by
// version number.
type StrategyRegistry struct {
	mu         sync.RWMutex
	strategies map[string]map[int]*DeterministicStrategy // nodeID → version → strategy
}

// NewStrategyRegistry creates an empty registry.
func NewStrategyRegistry() *StrategyRegistry {
	return &StrategyRegistry{
		strategies: make(map[string]map[int]*DeterministicStrategy),
	}
}

// Register adds a compiled strategy to the registry.
func (r *StrategyRegistry) Register(s *DeterministicStrategy) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.strategies[s.NodeID] == nil {
		r.strategies[s.NodeID] = make(map[int]*DeterministicStrategy)
	}
	r.strategies[s.NodeID][s.Version] = s
}

// Get retrieves a strategy by node ID and version; returns nil if not found.
func (r *StrategyRegistry) Get(nodeID string, version int) *DeterministicStrategy {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.strategies[nodeID][version]
}

// GetCurrent returns the highest-versioned strategy for a node.
func (r *StrategyRegistry) GetCurrent(nodeID string) *DeterministicStrategy {
	r.mu.RLock()
	defer r.mu.RUnlock()
	versions := r.strategies[nodeID]
	if len(versions) == 0 {
		return nil
	}
	latest := -1
	for v := range versions {
		if v > latest {
			latest = v
		}
	}
	return versions[latest]
}

// ListVersions returns sorted version numbers for a node.
func (r *StrategyRegistry) ListVersions(nodeID string) []int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	versions := r.strategies[nodeID]
	out := make([]int, 0, len(versions))
	for v := range versions {
		out = append(out, v)
	}
	// insertion sort (version lists are short)
	for i := 1; i < len(out); i++ {
		for j := i; j > 0 && out[j] < out[j-1]; j-- {
			out[j], out[j-1] = out[j-1], out[j]
		}
	}
	return out
}

// NextVersion returns the next monotonically increasing version for a node.
func (r *StrategyRegistry) NextVersion(nodeID string) int {
	versions := r.ListVersions(nodeID)
	if len(versions) == 0 {
		return 1
	}
	return versions[len(versions)-1] + 1
}

// ---------------------------------------------------------------------------
// NodeSnapshot — what DeterministicCompiler reads from a node
// ---------------------------------------------------------------------------

// NodeSnapshot is the interface that nodes must implement so the compiler can
// extract parameters and rules without depending on any concrete node type.
type NodeSnapshot interface {
	// NodeID returns the node's unique identifier.
	NodeID() string
	// NodeVersion returns the node's current version string.
	NodeVersion() string
	// Parameters returns a deep copy of the node's current parameters.
	Parameters() map[string]any
	// Rules returns the node's explicit rules, or nil to trigger synthesis.
	Rules() []StrategyRule
}

// ---------------------------------------------------------------------------
// DeterministicCompiler
// ---------------------------------------------------------------------------

// DeterministicCompiler freezes a node's current parameters into a versioned
// DeterministicStrategy.  Each compilation is assigned a monotonically
// increasing version number.  Any previous version can be retrieved via
// Rollback().
type DeterministicCompiler struct {
	Registry *StrategyRegistry
}

// NewDeterministicCompiler creates a compiler with a fresh registry.
func NewDeterministicCompiler() *DeterministicCompiler {
	return &DeterministicCompiler{Registry: NewStrategyRegistry()}
}

// NewDeterministicCompilerWithRegistry creates a compiler using an existing registry.
func NewDeterministicCompilerWithRegistry(r *StrategyRegistry) *DeterministicCompiler {
	return &DeterministicCompiler{Registry: r}
}

// Compile freezes node's current state into a DeterministicStrategy, registers
// it, and returns the result.
func (c *DeterministicCompiler) Compile(node NodeSnapshot) *DeterministicStrategy {
	params := node.Parameters()
	rules := node.Rules()
	if len(rules) == 0 {
		rules = synthesiseRules(params)
	}

	version := c.Registry.NextVersion(node.NodeID())
	checksum := computeChecksum(params)

	s := &DeterministicStrategy{
		StrategyID:  uuid.New().String(),
		NodeID:      node.NodeID(),
		NodeVersion: node.NodeVersion(),
		Version:     version,
		Parameters:  params,
		Rules:       rules,
		CompiledAt:  time.Now(),
		Checksum:    checksum,
	}
	c.Registry.Register(s)
	slog.Info("compiled strategy",
		"node_id", s.NodeID,
		"version", s.Version,
		"rules", len(s.Rules),
		"checksum", s.Checksum[:8],
	)
	return s
}

// Rollback retrieves a previously compiled strategy by version.
// Returns nil if the version does not exist.  Does NOT apply the strategy to
// the node; callers must do that explicitly.
func (c *DeterministicCompiler) Rollback(nodeID string, version int) *DeterministicStrategy {
	return c.Registry.Get(nodeID, version)
}

// GetCurrent returns the most recently compiled strategy for a node.
func (c *DeterministicCompiler) GetCurrent(nodeID string) *DeterministicStrategy {
	return c.Registry.GetCurrent(nodeID)
}

// ---------------------------------------------------------------------------
// Rule synthesis
// ---------------------------------------------------------------------------

// synthesiseRules builds a rule set from well-known parameter keys when the
// node does not provide explicit rules.
func synthesiseRules(params map[string]any) []StrategyRule {
	var rules []StrategyRule

	// RSI-style rules
	if v, ok := floatParam(params, "rsi_oversold"); ok {
		rules = append(rules, StrategyRule{
			Name:       "rsi_oversold_long",
			Direction:  DirectionLong,
			Conditions: []RuleCondition{{Field: "rsi", Op: "<=", Value: v}},
			Confidence: 0.7,
		})
	}
	if v, ok := floatParam(params, "rsi_overbought"); ok {
		rules = append(rules, StrategyRule{
			Name:       "rsi_overbought_short",
			Direction:  DirectionShort,
			Conditions: []RuleCondition{{Field: "rsi", Op: ">=", Value: v}},
			Confidence: 0.7,
		})
	}

	// Generic threshold rules
	sigConf, _ := floatParamDefault(params, "signal_confidence", 0.6)
	if v, ok := floatParam(params, "signal_threshold_long"); ok {
		rules = append(rules, StrategyRule{
			Name:       "signal_long",
			Direction:  DirectionLong,
			Conditions: []RuleCondition{{Field: "signal", Op: ">=", Value: v}},
			Confidence: sigConf,
		})
	}
	if v, ok := floatParam(params, "signal_threshold_short"); ok {
		rules = append(rules, StrategyRule{
			Name:       "signal_short",
			Direction:  DirectionShort,
			Conditions: []RuleCondition{{Field: "signal", Op: "<=", Value: v}},
			Confidence: sigConf,
		})
	}

	return rules
}

func floatParam(params map[string]any, key string) (float64, bool) {
	v, ok := params[key]
	if !ok {
		return 0, false
	}
	switch f := v.(type) {
	case float64:
		return f, true
	case float32:
		return float64(f), true
	case int:
		return float64(f), true
	}
	return 0, false
}

func floatParamDefault(params map[string]any, key string, def float64) (float64, bool) {
	if v, ok := floatParam(params, key); ok {
		return v, true
	}
	return def, false
}

// ---------------------------------------------------------------------------
// Checksum
// ---------------------------------------------------------------------------

func computeChecksum(params map[string]any) string {
	b, err := json.Marshal(sortedMap(params))
	if err != nil {
		return ""
	}
	sum := sha256.Sum256(b)
	return fmt.Sprintf("%x", sum)
}

// sortedMap produces a deterministic JSON representation of an arbitrary map.
func sortedMap(m map[string]any) map[string]any {
	// json.Marshal already sorts map keys, so just return m.
	// Nested maps are handled by encoding/json's natural key sorting.
	return m
}

// ---------------------------------------------------------------------------
// ValidationError (shared with memory_consolidation.go)
// ---------------------------------------------------------------------------

// ValidationError is returned when configuration parameters fail validation.
type ValidationError struct {
	Field   string
	Message string
}

func (e *ValidationError) Error() string {
	return fmt.Sprintf("validation error: field %q — %s", e.Field, e.Message)
}
